File: DynamicKey/AgoraDynamicKey/rust/src/stt_token_builder.rs
use crate::access_token;
use crate::rtc_token_builder::{Role, ROLE_PUBLISHER};

/// Builds a Token007 that carries RTC, RTM, and STT services.
pub fn build_token(
    app_id: &str,
    app_certificate: &str,
    channel_name: &str,
    rtc_account: &str,
    rtc_role: Role,
    rtc_token_expire: u32,
    join_channel_privilege_expire: u32,
    pub_audio_privilege_expire: u32,
    pub_video_privilege_expire: u32,
    pub_data_stream_privilege_expire: u32,
    rtm_user_id: &str,
    rtm_token_expire: u32,
) -> Result<String, Box<dyn std::error::Error>> {
    let mut token = access_token::new_access_token(app_id, app_certificate, rtc_token_expire);

    let mut service_rtc = access_token::new_service_rtc(channel_name, rtc_account);
    service_rtc.service.add_privilege(access_token::PRIVILEGE_JOIN_CHANNEL, join_channel_privilege_expire);
    if rtc_role == ROLE_PUBLISHER {
        service_rtc.service.add_privilege(access_token::PRIVILEGE_PUBLISH_AUDIO_STREAM, pub_audio_privilege_expire);
        service_rtc.service.add_privilege(access_token::PRIVILEGE_PUBLISH_VIDEO_STREAM, pub_video_privilege_expire);
        service_rtc.service.add_privilege(access_token::PRIVILEGE_PUBLISH_DATA_STREAM, pub_data_stream_privilege_expire);
    }
    token.add_service(Box::new(service_rtc));

    let mut service_rtm = access_token::new_service_rtm(rtm_user_id);
    service_rtm.service.add_privilege(access_token::PRIVILEGE_LOGIN, rtm_token_expire);
    token.add_service(Box::new(service_rtm));

    token.add_service(Box::new(access_token::new_service_stt()));

    token.build()
}

File: DynamicKey/AgoraDynamicKey/lua/src/stt_token_builder.lua
local access_token = require("agora_token.access_token")
local rtc_token_builder = require("agora_token.rtc_token_builder")

-- Builds a Token007 that carries RTC, RTM, and STT services.
-- rtc_token_expire is the whole token expiration. The whole token is invalid after this time.
-- RTC privilege expiration values and rtm_token_expire must not exceed rtc_token_expire.
local function build_token(app_id, app_certificate, channel_name, rtc_account, rtc_role, rtc_token_expire,
                           join_channel_privilege_expire, pub_audio_privilege_expire, pub_video_privilege_expire,
                           pub_data_stream_privilege_expire, rtm_user_id, rtm_token_expire)
    local token = access_token.new_access_token(app_id, app_certificate, rtc_token_expire)

    local service_rtc = access_token.new_service_rtc(channel_name, rtc_account)
    service_rtc.service:add_privilege(access_token.PRIVILEGE_JOIN_CHANNEL, join_channel_privilege_expire)
    if rtc_role == rtc_token_builder.ROLE_PUBLISHER then
        service_rtc.service:add_privilege(access_token.PRIVILEGE_PUBLISH_AUDIO_STREAM, pub_audio_privilege_expire)
        service_rtc.service:add_privilege(access_token.PRIVILEGE_PUBLISH_VIDEO_STREAM, pub_video_privilege_expire)
        service_rtc.service:add_privilege(access_token.PRIVILEGE_PUBLISH_DATA_STREAM, pub_data_stream_privilege_expire)
    end
    token:add_service(service_rtc)

    local service_rtm = access_token.new_service_rtm(rtm_user_id)
    service_rtm.service:add_privilege(access_token.PRIVILEGE_LOGIN, rtm_token_expire)
    token:add_service(service_rtm)

    token:add_service(access_token.new_service_stt())

    return token:build()
end

return {
    build_token = build_token,
}
